### PR TITLE
raftstore: prevent unsafe local read during merging (#8056)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #8056 and #7492 to release-3.0

---

### What problem does this PR solve?

Issue Number: close https://github.com/tikv/tikv/issues/8034

Problem Summary:

In this PR(https://github.com/tikv/tikv/pull/3116), it prevents unsafe local read by suspecting the leader lease.
https://github.com/tikv/tikv/blob/5755b3e67e910486fcd8ae76084e9979eec2bcd3/components/raftstore/src/store/peer.rs#L1456-L1466
But this may be too late to prevent unsafe local read.

After getting a raft ready, the msg will be sent immediately if it's a leader, which means the followers will know the `PrepareMerge` log is committed sooner. If a follower knows it, it will apply `PrepareMerge` log and propose `CommitMerge` to the target region, the merge process will succeed and can provide services.
In the previous fix, we suspect a leader lease after persisting the ready data. Between this time interval, local read will work as normal but it may get stale data.

I construct a simple test to reproduce this bug.

1. put (k1, v)
2. split region
2. set failpoint(pause) before suspecting the leader lease
3. merge left region to right region(only in other two pair regions, the leader of left region is stuck due to failpoint)
4. put (k1, v1)
5. read to the leader of left region => get (k1, v)

### What is changed and how it works?

What's Changed:

Suspect the leader lease before sending msg to other followers if a `PrepareMerge` log is committed.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

No.

### Release note 
* Fix an issue that read may get stale data during merging

### CODE REVIEW

File: src/raftstore/store/fsm/peer.rs [hunk 1]
```diff
@@ -1338,6 +1338,9 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         // We can't destroy a peer which is applying snapshot.
         assert!(!self.fsm.peer.is_applying_snapshot());
 
+        // Mark itself as pending_remove
+        self.fsm.peer.pending_remove = true;
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs:1342]: Why?

Comment (gengliqi) [reply src/raftstore/store/fsm/peer.rs:1342]: Because this PR also cherry-pick #7492

Comment (BusyJay) [reply src/raftstore/store/fsm/peer.rs:1342]: I recall master has changed the destroy flow and 3.0 doesn't have them, will it be a problem?

**Pull request merged**